### ISSUE TITLE
Reunification

### ISSUE THREAD

**Issue by bnmnetp** [skulpt/skulpt#825]
Hi All,

I've had a couple of days to catch my breath after last weeks SIGCSE conference.  Easy for me to say since it was in my new hometown of Minneapolis.  I had a few different conversations with people about skulpt in the hallways and lobbies around the conference.  They all went something like this....

Hey Brad, whats going on with Skulpt?  Are you still maintaining it?  What about better Python3 support? etc...

So,  yeah, this discussion is my attempt to try to get some conversation going about how we might re-unify our efforts to make Skulpt a great resource for the entire community.  Progress is being made on Skulpt, but it seems to be happening on other Forks and then staying on those forks rather than migrating back.  So I think we have gotten into a bad state where folks are reluctant to contribute to Skulpt because its clear that development has slowed here but continued in other places which is not only confusing, but IMHO kind of goes against the grain of being Open Source.

Between all of our projects Skulpt is doing amazing things to help many thousands of people learn to code in Python, and I would really love to see this continue.  But I really believe we need to re-unify our efforts if we are going to make serious progress.  After next year when Python2 comes to the end of its life I think we need a better Python3 story than we have today.

One solution is for me and/or others to manually create Pull Requests from known forks that have good stuff that we want.  The other is to have people working on other forks push their cool changes back to master.  Then I think we can unify our efforts to have a really stable Python3.

I Would love to hear the thoughts of others on how we might make this work.

Thanks,

Brad

Tagging:  @eah13 @albertjan @ebertmi @rixner @acbart @meredydd  for discussion and ideas.   Please feel free to tag others that may have ideas too.

**Comment by rixner** [comment]
I have not kept up with other forks, so I do not know what people have been doing that has not made its way back here.  I agree with @bnmnetp wholeheartedly, though, that it would be great to bring the vibrancy back into the project.

I have come to the conclusion that for my purposes, all my future development will be on Python 3.  I think that Skulpt risks total irrelevance if we don't shore up Python 3 support and have a very good story there.

To that end, I have been thinking about where to start, and one thing that I've found prohibitive is that the grammar is neither updated nor easy to update (and I think it relies on an older version of pgen, that CPython has since abandoned).  I feel like we should be able to leverage existing work for that, though.  I was able to use [antlr](https://www.antlr.org/), for instance, to build a parser for Python 3.6 with somewhat limited effort.  Tying it in to Skulpt is another thing, but at least then we would have a complete grammar upon which we can build.

But, I think an even better starting point is to galvanize the community into bringing back all of the great work that has already been done.  I am fine if we just go look for it and create our own pull requests, but it seems like perhaps the first step would be to simply contact those with active forks and invite them (back) into the fold?

I also suspect that if we updated our workflow (to perhaps use node, npm, webpack, etc.) we might attract more developers...  But, we really should get as much pulled back into the project as we can before making such a major change.

**Comment by acbart** [comment]
Besides harrassing Brad at SIGCSE (sorry again!), I'm certainly someone responsible for forking and then not contributing back. Just off the cuff, the BlockPy project has created:

* An AST module
* A custom MatPlotLib module with histograms, line plots, and scatter plots
* Our own Requests and File Input mechanisms

And I'm sure I'm forgetting a lot of little details of our work, plus WiP efforts. Lately, for example, I've been experimenting with a system for transliterating Python programs to Skulpt code, in order to improve performance without needing a full compilation.

However, we are facing a split in the road of how we want to move forward with BlockPy's execution engine. We are committed to moving towards Python 3, and I imagine that one way or the other we'll need to move to it in the next year. I have been looking at Brython as an alternative, which has its own tradeoffs. I think I would still prefer to work with Skulpt, if that's a possibility.

Certainly modernizing the workflow would encourage me further. A major, deal-making goal for me would be to get full compatibility with our [Pedal](https://github.com/acbart/pedal) library (for analysis of student Python programs to provide feedback). This would require a functioning version of `exec`, better support for kwargs, and a few other features.

Interestingly, last week I tasked an undergraduate capstone team from UD with the problem of writing a Python 3 compatible ANTLR parser with a JS target. In our case, they are focused on improving the dual block/text interface for BlockPy. It sounds like you are already making progress on this route. I don't know how my team could be supportive of that effort, but at least it sounds like we are sharing a vision in that regard.

A concern from my perspective is how much of this has to be a rewrite-from-scratch vs. incremental development. I know there are some holes in the Python compatibility that frustrate emulating certain features. From my perspective as someone interested but unsure about where to apply our time, energy, and resources, it would be nice to get a sense of what Skulpt is currently lacking and what it will take to get full Python 3 support going. I suppose I'm asking for an [updated roadmap](https://github.com/skulpt/skulpt/issues/400), and for a clear vision/argument of how [Skulpt differentiates itself from Brython](https://github.com/skulpt/skulpt/issues/825).

**Comment by albertjan** [comment]
Hi! 👋 

I'll make a point of merging some of trinkets stuff back into the main repo as soon as somebody merges the super PR :smile:.

On python3 I would like to point you to this branch: https://github.com/trinketapp/skulpt/tree/py3k which contains a tokeniser -> parser -> ast that understand py3 code. This is a hard fork however we lose the py2 parser. Although it's probably quite easy to make a `from __past__ import print_statement` or something along those lines. 

A complete rewrite would not make sense to me really, not without someone working on this nearly fulltime for a good while. 😄 

About the workflow I agree with @rixner there. I have played a little with getting skulpt to run on node, but because the file combining system we use and that of node don't match at all that involved a lot of unsavoury hacks. I think a good first step in this direction would be to update to the latest google closure compiler. And then use javascript imports and let closure compiler spit out a compiled result, instead of us concatenating the files.

Oh and while you're looking at PR's have a look at this one too: https://github.com/skulpt/skulpt/pull/736

**Comment by bnmnetp** [comment]
Thanks for the response @albertjan --- I will merge #694 and take a look at #736 Both seem well worth getting on the main branch, even if others have suggestions for further improvements.

I would like to delegate someone else to take primary responsibility for Python3 work.  I have a lot of other stuff on my plate and In all honesty there are others better qualified for this rather large undertaking.

My thoughts on python3 are that we ought to make a list of what we think is essential to claim some kind of Python3 compatibility.  I'm in agreement that a complete rewrite seems like more work than is probably necessary.  For my intro courses we pretty much have everything I think is important.  I've been easily migrating from skulpt in the browser to Python3 in an IDE at mid-semester with no ill effects.  But as my use cases expand to more upper level courses I think its likely there are other things that will be required/make sense.

Brad

**Comment by acbart** [comment]
I think the biggest Python 3 thing for me is the type annotations. Even if it's just parser support, that'd be a big win. I want to see intro CS1 curriculum take advantage of static typing, both for pedagogical benefits and to shut up the people who don't understand the difference between static/dynamic and strong/weak typing.

I'll put in another plea for `exec`, I suppose. I know it's not really a Python 3 thing, but it's definitely a big, indigestible thing on my plate of issues.

**Comment by albertjan** [comment]
Well here it is folks! #827 Please lets not let this linger 😄 rebasing / merging this will not be fun!

**Comment by bnmnetp** [comment]
A python3 feature on my wishlist would be f-strings

**Comment by bnmnetp** [comment]
@acbart I would love it if you could PR back your work on the requests module.  How do you handle CORS issues with your implementation?

**Comment by bnmnetp** [comment]
> Well here it is folks! #827 Please lets not let this linger 😄 rebasing / merging this will not be fun!

Your wish has been granted 🎉

**Comment by bnmnetp** [comment]
Now if I can just remember how to update skulpt.org!!   I think I need to remember how to work Google AppEngine!!

**Comment by acbart** [comment]
> @acbart I would love it if you could PR back your work on the requests module. How do you handle CORS issues with your implementation?

Hm, I'm not sure you actually want it. I had forgotten the details until now, but basically we just reused the urllib module to make the subset of Requests that we needed (really just `get` and the `json` attribute). Our contribution is trivial, all things considered - in fact, it's weirder than that.  Our version allows you to specify a function (`Sk.requestsGet`) that emulates the actual call by passing it along to our server, which needs to have a hashed filename that matches the passed in URL. That way we avoid the CORS issue. It's... hacky. Sorry to oversell :/

**Comment by albertjan** [comment]
@bnmnetp I think that happens automatically when we do a new release. We do download gae in dist-update. 😊

**Comment by bnmnetp** [comment]
Oh right!  That should happen automatically when we do a new tag.

Unfortunately the dist-update script is a little behind the times.  gae is now part of Google Cloud and the wget 404's when I tried to grab an older version.

**Comment by rixner** [comment]
Like @acbart, I'd also like to see type annotations.  I still do think there is value in having a complete grammar, just with unimplemented functionality, so that one could add functionality without also having to figure out how to update the grammar.  Especially if we could sync it with CPython's grammar files or another project that keeps the grammar up to date.  But, I get it if that's too much reimplementation effort.

I'd like to be able to have code with type annotations run through Skulpt properly so that the same code that runs in Skulpt could also be externally run through a static type checker like mypy.

**Comment by meredydd** [comment]
+1 to that, @rixner - a complete grammar in the parser would be a wonderful thing, for three reasons:

1. It means we can parse everything, and produce clear errors on unsupported things in the compiler ("we should support this, but we don't")
2. It will mean Anvil's autocompleter will no longer break the moment someone uses an f-string in a module
3. It means that those of us familiar with the compiler but slightly afraid of the parser (_ahem_) can start getting on with incrementally implementing Python 3 features -- right now I wouldn't know where to start.

I'm not sure what we do about the hard fork problem - any ideas, @albertjan? Could we just hack the grammar to support the `print` statement, and everything else is a superset of python 2?

From my (Anvil's) point of view, Skulpt's Python 2 support is sufficiently stable, and the Python 2 sunset is now sufficiently close, that I'm willing to jump ship and say "Skulpt now has a python-3-ish main version and a python-2 legacy fork" (the python 2 support is sufficiently stable for our needs that I'd be happy to let it more or less rot in place). It's a dangerous leap, though -- I understand how the project could get caught between the two stools.

**Comment by acbart** [comment]
It might not be time to talk about this feature request, but one thing we did in BlockPy was hack on `endlineno` and `col_endoffset` to the [parser](https://github.com/RealTimeWeb/skulpt/blob/master/gen/astnodes.js#L378) - this was necessary for some of the futzing we do with the block/text conversion, as I recall. Might be useful for others to have such a feature, depending on what folks are working on.

**Comment by rixner** [comment]
I'm ready to just move to Python 3.  As @meredydd said, I feel the Python 2 support is sufficiently stable for my needs.  So, I'm an advocate of just tagging a stable Python 2 point and moving all development to Python 3 and no longer putting effort into making sure Skulpt continues to support Python 2.

I know @meredydd is +1'ing my proposal, but I really want to reiterate that the current parser is really difficult to deal with, is incomplete, and does not keep up with Python.  If we want Skulpt to maintain relevance, I think we would be in a much better place if we could easily update the grammar with each Python release, and have everything that is not yet implemented fail gracefully.

Right now, when one wants to add a new feature that requires a grammar change, you look at the parser mess and think "is it worth it?"  That doesn't seem like a good place for the project. If we want to keep the parser, can someone who really understands it document the process for actually updating the grammar at least?

**Comment by rixner** [comment]
@albertjan is your py3k branch using the full Python 3.7 grammar?  I just looked and it seems to be included in there...  Or at least something that looks awfully close (Grammar.txt).

**Comment by meredydd** [comment]
Sorry, I realise my last paragraph ended up saying something I didn't mean - my apologies. I am emphatically *not* in favour of an actual fork that leaves the Python 2 codebase in the dust.

Anvil users have code in production that relies on Python 2. We are going to have to carry on supporting them for some time, which means we are going to have to carry on maintaining bindings to a version of Skulpt that supports Python 2. If the Python 2 and Python 3 versions of Skulpt diverge in their ABIs, we will probably end up stranded on the Python 2 branch (or effectively forking Anvil, which would be deeply painful).

I would strongly advocate maintaining the current level of Python 2 support in Skulpt as we move forward. Ideally I'm imagining two grammars that yield the same set of AST objects, and a compiler with feature flags (I think the current `Sk.__future__` is an excellent template for this).

By "rot in place" I mean that I'd be happy if we ended up eventually concentrating all feature development to Python 3, and doing no more work on Python 2.7 compatibility than is necessary to avoid breaking the tests we already have.

**Comment by rixner** [comment]
@meredydd , don't you think that's going to become increasingly difficult and problematic for a version of the language that is about to die...  Right now, we have pretty decent Python 2 and hacked up Python 3 on top of it.  If we go the route you are suggesting, we need to make it pretty decent Python 3 and then hacked up Python 2 on top of that?  How much enthusiasm are people going to have for that once Python 2 is EOL?  For instance, if there is a pull request that dramatically improves Python 3, but breaks Python 2, then what?

Also, it seems like we may already be at that point, as it sounds like @albertjan has already made a lot of Python 3 headway, but it is incompatible with the Python 2 infrastructure (unless I misunderstood his comment above).  If we can find a way to keep them together that does not slow down Python 3 development, I certainly am fine with that.  If not, though, I think we have to do what is necessary to speed up Python 3 development.

Please don't get me wrong, I completely sympathize with your situation!  I am going to be supporting Python 2 until the end of time, as well.  I just decided a long time ago to keep them completely separate.  I get how that may not be a viable option for you!

**Comment by csev** [comment]
Hi all,

TL;DR;  I have not lifted a finger to help this project for a number of years - so this is only opinion and nothing more.

I am a super-happy customer of Skulpt in my Python MOOC on Coursera and Python for Everybody textbook and www.py4e.com <http://www.py4e.com/>.  I last upgraded my skulpt.js 18-Nov-2016.  

I built a complete Python 2.0 course/book and then re-did the course/book in "Python 3.0" - still in Skulpt.  I definitely owe the Skulptors dinner for the success that has come my way because of Skulpt.  Thanks.

That said, I have been watching this reunification conversation with interest because I think that Python in the browser might someday become a far bigger thing than what we already have.   Skulpt either will be a core technology or a critical stepping stone into a cool future.  I really like hearing of the tweaks that let Python talk to the DOM.  I don’t plan to use any of that - but who knows - someday :)

I think it is far past time to put all the Python 2.0 (plus some 3.0) work on a branch and focus on an awesome 3.0 only product going forward.  This does not mean that the 2.x stuff will rot - it is very mature and will get effort as folks need it.   It has solid tests and should not have regressions even if a little forward movement happens.

We have been able to walk this fine line of 2.x and a little 3.x for a while - but it will become more difficulty as 3.x progresses and we want an increasingly high fidelity 3.x in Skulpt. Things like type checking, regular expression processing, far richer stdlibs, etc etc.  The underlying design of Python 3.0 is to put in changes that break 2.0.  Skulpt just has not gotten far enough into deep and complete Python 3.0 support to find the scary places where 2.x and 3.x will be in conflict with one another.

My feeling (and it is only a feeling) is that by trying to fully support both 2.x and 3.x work for ever in a single code base will lead to a poor 3.x implementation.  If there are elements like parsers or code generators or optimizers that can be shared without compromising - sure - but my instinct is that this will become effort that does not justify value.

Thanks for listening.   I don’t have to do any of the work so I will just go back to lurking.

/Chuck Severance
University of Michigan School of Information

**Comment by meredydd** [comment]
_rixner:_
> Right now, we have pretty decent Python 2 and hacked up Python 3 on top of it. If we go the route you are suggesting, we need to make it pretty decent Python 3 and then hacked up Python 2 on top of that? 
Yep, that's pretty much exactly what I'm suggesting.

> How much enthusiasm are people going to have for that once Python 2 is EOL?
Hopefully very little, and we can abandon our Python 2 support quickly! (I'm not actually planning to support Python 2 in Anvil forever.) But Python 2 isn't EOL for another year, and people have code in production using Skulpt in Python 2 mode. Until we've given them a chance to migrate, we can't really cut them off. And in that first version that supports decent Python 3, we're going to have to let people edit and run their Python 2 programs as they migrate -- and for us at least, that would become _extremely_ painful if we effectively had to fork our entire platform to track diverging Javascript ABIs.

_csev:_
> The underlying design of Python 3.0 is to put in changes that break 2.0.
While this is true, the change that forced the split (and pretty much the only long-term compatibility issue) is in the `str` type -- which is sufficiently patchily supported in our current version (it's basically a JS string) that the backward-compatibility could get away with being very hacky indeed! 

I appreciate that running code beats arguments, and if we weren't tooling up for PyCon SK this weekend I would currently be checking out @albertjan's py3k branch, hacking a `print` statement into the grammar, and seeing how much of the Python 2 tests (and/or our customers' code) still parse.

**Comment by csev** [comment]
Now I am drifting off topic.  Anvil is really really cool.

Don’t take this the wrong way.   Reminds me of Visual Basic 4.3 and RealBasic - Two of my favorite IDEs.   Except that Anvil is client/server.

Really cool.

**Comment by albertjan** [comment]
@meredydd adding a print statement to the grammar is simple enough.

**Comment by albertjan** [comment]
@rixner it's this version I think https://github.com/python/cpython/blob/52c4e7cc84702750bb75d5423da01d01bcdfdf39/Grammar/Grammar I wonder what has changed 😄

**Comment by bnmnetp** [comment]
@albertjan -- Thanks again for this work.  I'm wondering what you found to be problematic with the Sk.externalLibraries implementation that you pulled it out.  It seemed to work well enough for us.  I'm not sure its that big of a deal as I think I would rather just add custom libraries to the Runestone Fork...  Just wondering.

**Comment by albertjan** [comment]
Two reasons really:

1. (practical) it was using a synchronous xhr request, which will be disallowed some time in the future. And making it use a non sync request and suspensions would have made the already quite intricate import.js even more difficult. And since @meredydd had made it possible to do async `Sk.read` implementations and I made `Sk.onBeforeImport` and `Sk.onAfterImport` also handle async calls.

2. (philosophical) I/We felt it wasn't really skulpts responsibility to implement this intricate dependency loading mechanism. And if we decided it was a responsibility it should be part of a out of the box implementation of read.

I could make an example of how to implement it like this if you want.

**Comment by meredydd** [comment]
Update: @albertjan and I have got Skulpt working, and passing all our existing (Python 2 and 3) tests, on top of the Python 3.7.3 grammar. This is being tracked in pull request #905.
